bentoml/yatai/client/bento_repository_api.py

```diff
@@ -51,13 +57,74 @@
     safe_retrieve,
     load_from_dir,
 )
+from bentoml.yatai.proto.yatai_service_pb2_grpc import YataiStub
 from bentoml.yatai.status import Status
 
 
 logger = logging.getLogger(__name__)
 yatai_proto = LazyLoader('yatai_proto', globals(), 'bentoml.yatai.proto')
 
 
+class BentoUploadStreamRequests:
+    """
+    A class for iterating over a file to generate upload bento requests
+
+    Args:
+        bento_name: str
+        bento_version: str
+        file_path: path
+        chunk_size optional: int
+    """
+
+    def __init__(
+        self, bento_name, bento_version, file_path,
+    ):
+        self.bento_name = bento_name
+        self.bento_version = bento_version
+        self.bundle = open(file_path, 'rb')
+        (
+            self.bundle_size,
+            self.bundle_chunk_count,
+            self.chunk_size,
+        ) = get_file_size_and_chunk_count(file_path)
+        self.sent_chunk_count = 0
+        self.file_index = 0
+        self.sent_init_message = False
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        if self.sent_chunk_count == 0 and self.sent_init_message is False:
+            request = UploadBentoRequest(
+                bento_name=self.bento_name, bento_version=self.bento_version,
+            )
+            self.sent_init_message = True
+            return request
+        elif self.sent_chunk_count < self.bundle_chunk_count:
+            current_file_end = min(self.bundle_size, self.file_index + self.chunk_size)
+            self.bundle.seek(self.file_index)
+            chunk = self.bundle.read(self.chunk_size)
+            self.file_index = current_file_end
+            request = UploadBentoRequest(
+                bento_name=self.bento_name,
+                bento_version=self.bento_version,
+                bento_bundle=chunk,
+            )
+            self.sent_chunk_count += 1
+            return request
+        else:
+            self.bundle.close()
+            raise StopIteration
+
+
+def process_grpc_error(grpc_error):
+    if grpc_error.code() == grpc.StatusCode.DEADLINE_EXCEEDED:
```

Comment:
why special case this error type?  what's the output of `grpc_error.details()` when `grpc_error.code() == grpc.StatusCode.DEADLINE_EXCEEDED`?